lldb/source/Plugins/TypeSystem/Swift/LRUCache.h

```diff
@@ -0,0 +1,98 @@
+//===-- LRUCache.h ----------------------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef liblldb_LRUCache_h_
+#define liblldb_LRUCache_h_
+
+#include "llvm/ADT/STLFunctionalExtras.h"
+#include <list>
+#include <unordered_map>
+
+namespace lldb_private {
+namespace swift_demangle {
+
+template <typename Key, typename Value> class LRUCache {
+public:
+  LRUCache(size_t capacity) : capacity_(capacity) {}
+
+  std::optional<Value> get(const Key &key) {
+    auto it = map_.find(key);
+    if (it == map_.end())
+      return std::nullopt;
+    list_.splice(list_.begin(), list_, it->second);
+    return it->second->second;
+  }
+
+  void put(const Key &key, const Value &value) {
+    auto it = map_.find(key);
+    if (it != map_.end()) {
+      it->second->second = value;
+      list_.splice(list_.begin(), list_, it->second);
+    } else {
+      if (capacity_ > 0) {
+        if (list_.size() >= capacity_) {
+          auto last = list_.back();
+          map_.erase(last.first);
+          list_.pop_back();
+        }
+        list_.emplace_front(key, value);
+        map_[key] = list_.begin();
+      }
+    }
+  }
+
+  Value get_or_create(const Key &key, llvm::function_ref<Value()> factory) {
+    if (auto it = map_.find(key); it != map_.end()) {
+      list_.splice(list_.begin(), list_, it->second);
+      return it->second->second;
+    }
+    const auto value = factory();
+    if (capacity_ > 0) {
+      if (list_.size() >= capacity_) {
+        map_.erase(list_.back().first);
+        list_.pop_back();
+      }
+      list_.emplace_front(key, value);
+      map_[key] = list_.begin();
+    }
+    return value;
+  }
+
+  Value get_or_create(Key &&key, llvm::function_ref<Value()> factory) {
```

Comment:
I wonder if it's possible to share more code between the two versions of this function using forwarding references, but if it makes the code too complicated it might not be worth it.

Reply:
Yes, we could, but I switched to StringRef for the key, so we don't need this rvalue optimization anymore.